packages/astro/src/vite-plugin-env/index.ts

```diff
@@ -93,7 +104,15 @@ export default function envVitePlugin({ config: astroConfig }: EnvPluginOptions)
 			while ((match = pattern.exec(source))) {
 				const start = match.index;
 				const end = start + match[0].length;
-				const replacement = '' + replacements[match[1]];
+				let replacement = '' + replacements[match[1]];
+				// If we match exactly `import.meta.env`, define _only_ referenced private variables
+				if (match[0] === 'import.meta.env') {
+					replacement = `(Object.assign(import.meta.env,{`
+					for (const key of references.values()) {
+						replacement += `${key}:${privateEnv[key]},`
+					}
```

Comment:
Importantly, we only inline secret vars that we know are referenced in the file.